proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,332 @@
+      - Extractive QA
+      - Structured Data to Text
+      - Coreference Resolution
+      - Code Repair
+      - Code Error Generation
+      - Dialogue Context Generation
+      - Closed-Book QA
+      - Next Sentence Prediction
+      - Paraphrasing Identification
+      - Conversational Question Answering
+      - Topic Classification
+      - Mathematical QA
+      - Dialog Next Turn Prediction
+      - Grammatical Acceptability
+      - Punctuation fixing
+
+    [T0-SF]:
+      - Adversarial QA
+      - Question Generation
+      - Commonsense Reasoning
+      - Title Generation
+      - Dialogue Turn Prediction
+      - Predict Span Indices
+      - Context Generation
+
+    [NIV2]:
+      - Program Execution
+      - Text Matching
+      - Toxic Language Detection
+      - Cause Effect Classification
+      - Information Extraction
+      - Textual Entailment
+      - Wrong Candidate Generation
+      - Named Entity Recognition
+      - Commonsense Classification
+      - Fill-in-the-blank
+      - Text Completion
+      - Sentence Composition
+      - Question Understanding
+
+    [CoT Reasoning]:
+      - Explanation Generation
+      - Generate Question And Answer
+      - Grade School Math Word Problems
+      - Algebraic Question Answering
+      - Common Sense Reasoning Over Entities
+      - Common Sense Reasoning For QA
+      - Passage Based Question Answering
+      - Sense-Making And Explanation
+
+  # Drawbacks
+- One potential drawback of MultiTaskComponent is that it may require a significant amount of computational resources
+  to use. This may limit its use in applications or environments where there are constraints on the available hardware
+  or software resources. In addition, MultiTaskComponent may not be suitable for all NLP tasks or use cases, and users
+  may need to evaluate its performance and suitability for their specific needs and requirements.
+
+
+- During the prototyping phase we found that in order to follow current pipeline design patterns we needed to make
+  certain compromises. If a pipeline has more than one MultiTaskComponent, we can not simply use one instance of
+  MultiTaskComponent and switch the tasks in runtime. Instead, we need to create a new instance of MultiTaskComponent
+  for each task in the pipeline. This is a limitation of the current pipeline design and we hope to
+  address with invocation layer changes in the future.
+
+
+
+  # Alternatives
+
+- One alternative to MultiTaskComponent is to continue to use separate models for each NLP task in Haystack. This
+  can enable users to tailor the model to the specific requirements of each task, and to potentially improve the
+  performance of the model for that task by additional fine-tuning or model adaptation via GPL. However, using separate
+  models may require these complex and computationally intensive training and deployment processes, and may not be as
+  efficient or flexible as using a single, unified model.
```

Comment:
I think I don't fully understand how this is an alternative. 

It seems like `MultiTaskComponent` provides a nice `TaskTemplate` framework to be able to interact with generative models which is definitely missing in Haystack. And to me, that seems like one of the main value propositions of this proposal. And the alternative to that would be to require users to create custom nodes that would handle this kind of prompt engineering. 

Additionally, even with the introduction of `MultiTaskComponent` I could choose to use different generative models in my pipeline for different tasks so this wouldn't necessarily lead to having a single unified model. However, I do agree if that I could reuse the same `MultiTaskComponent` node within one pipeline but with a different `TaskTemplate` at run time that hardware resources could be saved.

Reply:
Next version 😁